orangecontrib/spectroscopy/io/neaspec.py

```diff
@@ -11,195 +11,285 @@
+            meta_cols = 4
+
+        Meta_data = np.zeros((int(N_rows), meta_cols), dtype="object")
+
+        # Filling up meta_data with positions, run and channelnames
         alpha = 0
         beta = 0
-        Ch_n = int(Channel.size)
 
-        for i in range(0, N_rows, Ch_n):
+        for _, i in enumerate(
+            range(0, Max_row * Max_col * N_chn * Max_run, N_chn * Max_run)
+        ):
             if beta == Max_row:
                 beta = 0
                 alpha = alpha + 1
-            Meta_data[i:i + Ch_n, 2] = Channel
-            Meta_data[i:i + Ch_n, 1] = alpha
-            Meta_data[i:i + Ch_n, 0] = beta
+
+            for jrun in range(Max_run):
+                Meta_data[
+                    i
+                    + (Max_row * Max_col * N_chn * jrun) : i
+                    + N_chn
+                    + (Max_row * Max_col * N_chn * jrun),
+                    -1,
+                ] = channelnames
+
+                if "Run" in list(data.keys()):
+                    Meta_data[
+                        i
+                        + (Max_row * Max_col * N_chn * jrun) : i
+                        + N_chn
+                        + (Max_row * Max_col * N_chn * jrun),
+                        -2,
+                    ] = jrun
+
+                Meta_data[
+                    i
+                    + (Max_row * Max_col * N_chn * jrun) : i
+                    + N_chn
+                    + (Max_row * Max_col * N_chn * jrun),
+                    1,
+                ] = ypos[int(alpha), int(beta)]
+
+                Meta_data[
+                    i
+                    + (Max_row * Max_col * N_chn * jrun) : i
+                    + N_chn
+                    + (Max_row * Max_col * N_chn * jrun),
+                    0,
+                ] = xpos[int(alpha), int(beta)]
+
             beta = beta + 1
 
-        waveN = file[0:int(Max_omega), 3]
-        metas = [Orange.data.ContinuousVariable.make("row"),
-                 Orange.data.ContinuousVariable.make("column"),
-                 Orange.data.StringVariable.make("channel")]
+        # Neaspec tends to name the independent variable differently all the time
+        # Try to prepare for all the names we know so far
+        if "Run" in list(data.keys()):
+            waveN = data["M"][0 : int(Max_omega)] * 1e6
+            metas = [
+                Orange.data.ContinuousVariable.make(MAP_X_VAR),
+                Orange.data.ContinuousVariable.make(MAP_Y_VAR),
+                Orange.data.ContinuousVariable.make("run"),
+                Orange.data.StringVariable.make("channel"),
+            ]
+        elif "Wavelength" in list(data.keys()):
+            waveN = data["Wavelength"][0 : int(Max_omega)]
+            metas = [
+                Orange.data.ContinuousVariable.make(MAP_X_VAR),
+                Orange.data.ContinuousVariable.make(MAP_Y_VAR),
+                Orange.data.StringVariable.make("channel"),
+            ]
+        else:
```

Comment:
Last two repeat a lot. Could be unified with a small exception.

Code here repeats a lot and could be shorter. At least first part of metas, with positions, can be created once before the ifs.